borsh/tests/common_macro.rs

```diff
@@ -0,0 +1,110 @@
+#[allow(unused)]
+macro_rules! set_insert_deser_assert_macro [
+
+    [$set: ident, $data: ident, $($key: expr),*] => [
+        $($set.insert($key));*
+        ;
+
+        let $data = $set.try_to_vec().unwrap();
+        #[cfg(feature = "std")]
+        insta::assert_debug_snapshot!($data);
+    ]
+];
+
+#[allow(unused)]
+macro_rules! map_insert_deser_assert_macro [
+
+    [$map: ident, $data: ident, $($key: expr => $value: expr),*] => [
+        $($map.insert($key, $value));*
+        ;
+        let $data = $map.try_to_vec().unwrap();
+        #[cfg(feature = "std")]
+        insta::assert_debug_snapshot!($data);
+    ]
+];
+
+#[allow(unused)]
+macro_rules! set_wrong_order_test [
+
+    [$test_name: ident, $set_type: ty] => [
+
+        #[test]
+        fn $test_name() {
+            let mut data = vec![];
+            let arr_key = ["various".to_string(), "foo".to_string(), "many".to_string()];
+            let len = arr_key.len() as u32;
+            u32::serialize(&len, &mut data).expect("no error");
+
+            for key in &arr_key {
+                key.serialize(&mut data).expect("no error");
+            }
+
+            let result = from_slice::<$set_type>(&data);
+
+            #[cfg(not(feature = "de_strict_order"))]
+            {
+                let result = result.unwrap();
+                assert_eq!(result.len(), arr_key.len());
+                for key in &arr_key {
+                    assert!(result.contains(key));
+
+                }
+
+            }
+
+            #[cfg(feature = "de_strict_order")]
+            {
+                assert!(result.is_err());
+
+                assert_eq!(result.unwrap_err().to_string(), ERROR_WRONG_ORDER_OF_KEYS);
+            }
```

Comment:
You implemented it well, but semantically this feels wrong to me. This test ensures that if canonicity is broken, we still read the data just fine.

Let me think out loud here:

Currently, we ask people to opt-in for strong consistency, and if any crate in the dependencies tree opts-in (when borsh is used in several crates using different features), the strong consistency checks will be enabled everywhere, and potentially erroring on incorrectly constructed payloads and affecting performance (due to validation on deserialization).

Alternatively, we may enable the checks by default and create a feature to opt-out ("non-strict-mode"), and the risk will be the opposite that someone enabling the feature in one crate will disable canonicity checks (on deserialization), and potentially speed things up.

I cannot say I like either of the approaches more than the other, but this test case looks like we endorse broken canonicity. Also, I feel that to be on the safe side of things, I would enable this feature by default (when making a breaking change release), but at the same time, it would introduce a performance hit onto the users who always construct and read the data with borsh-rs (e.g. near-sdk-rs smart contracts, nearcore internals), and serialization is canonical there.

@dj8yfo What do you think?

Reply:
I think the test just creates a piece of runnable documentation of actual behavior and doesn't attempt to endorse or denounce canonicity.
A feature by itself doesn't nudge crate's users in any particular direction, enabling/disabling it by default does. I think that decision at some point is up to you @frol. 
To better understand performance impact, a benchmark with some data type used in practice is due as part of this task/another pr. 
My intuition is that performance impact must be negligible for most practical types, which don't attempt to allocate memory on comparisons. But, as it often happens with intuition, it may be wrong.

Reply:
:high_brightness: